python/lsst/pipe/base/graph/graph.py

```diff
@@ -614,19 +637,75 @@ def save(self, file):
         file : `io.BufferedIOBase`
             File to write pickle data open in binary mode.
         """
-        pickle.dump(self, file)
+        buffer = self._buildSaveObject()
+        file.write(buffer.getvalue())
+
+    def _buildSaveObject(self) -> io.BytesIO:
+        # make some containers
+        pickleData = []
+        nodeMap = {}
+        taskDefMap = {}
+        protocol = 3
+
+        # counter for the number of bytes processed thus far
+        count = 0
+        # serialize out the task Defs recording the start and end bytes of each
+        # taskDef
+        for taskDef in self.taskGraph:
+            # compressing has very little impact on saving or load time, but
+            # a large impact on on disk size, so it is worth doing
+            dump = lzma.compress(pickle.dumps(taskDef, protocol=protocol))
+            taskDefMap[taskDef.label] = (count, count+len(dump))
+            count += len(dump)
+            pickleData.append(dump)
+
+        # serialize the nodes, recording the start and end bytes of each node
+        self_list = list(self)
+        for node in self_list:
+            node = copy.copy(node)
+            taskDef = node.taskDef
+            object.__setattr__(node, 'taskDef', taskDef.label)
+            # compressing has very little impact on saving or load time, but
+            # a large impact on on disk size, so it is worth doing
+            dump = lzma.compress(pickle.dumps(node, protocol=protocol))
+            pickleData.append(dump)
+            nodeMap[node.nodeId.number] = (count, count+len(dump))
+            count += len(dump)
+
+        # pickle the taskDef byte map
+        taskDef_pickle = pickle.dumps(taskDefMap, protocol=protocol)
+
+        # pickle the node byte map
+        map_pickle = pickle.dumps(nodeMap, protocol=protocol)
+
+        # record the sizes as 2 unsigned long long numbers for a total of 16
+        # bytes
+        map_lengths = struct.pack('>HQQ', SAVE_VERSION, len(taskDef_pickle), len(map_pickle))
+
+        # write each component of the save out in a deterministic order
+        buffer = io.BytesIO()
+        buffer.write(map_lengths)
+        buffer.write(taskDef_pickle)
+        buffer.write(map_pickle)
+        for dump in pickleData:
+            buffer.write(dump)
+        return buffer
```

Comment:
Are you talking about the fact that there is the bytes of the object and then it written out to the buffer, or are you taking about the data structures inside the nodes?

Reply:
I'm taking about `dump` having all bytes that you want to dump and then you copy those into `buffer` too. It's a temporary spike for sure but it can be quite big spike. One way to remedy that is to remove pieces from `dump` as you write them into `buffer` but it may not help much depending on how memory allocation works. Also filling `buffer` will likely to cause re-allocations and fragmentation too. I don't know how to avoid all this mess if you want to construct whole buffer in memory before writing (or if it's worth spending time trying to improve it).

Reply:
Its a bit of a chicken and egg here, you cant dump directly, as the ordering that things is written is important here, and `ButlerURI` does not support subsequent write in append mode for all of the backends. I think there are tricks I can play with the buffer though so I will look into it.

Reply:
I think I might be able to avoid some of the copies using bytearrays, I will try and do some size tests.